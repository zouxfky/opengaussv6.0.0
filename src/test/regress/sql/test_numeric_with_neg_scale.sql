create schema numeric_negative_scale_test;
set current_schema to numeric_negative_scale_test;

set behavior_compat_options = 'truncate_numeric_tail_zero';
-- test normal functions(valid values)
CREATE TABLE t1(a numeric(4,-3), b numeric(5,-2), c numeric(6, -1));
CREATE TABLE t2(a numeric(3,-4), b numeric(2,-5), c numeric(1, -6));
CREATE TABLE t3(a numeric(5, 7), b numeric(2, 4), c numeric(10, 12));
\d t1;
\d t2;
\d t3;

INSERT INTO t1 VALUES (1234567, 1234567, 1234567);
INSERT INTO t1 VALUES (123.9435435, 123.9435435, 123.9435435);
INSERT INTO t1 VALUES (0.8293453, 0.8293453, 0.8293453);
INSERT INTO t1 VALUES (-1234567, -1234567, -1234567);
INSERT INTO t1 VALUES (-123.9435435, -123.9435435, -123.9435435);
INSERT INTO t1 VALUES (-0.8293453, -0.8293453, -0.8293453);
INSERT INTO t1 VALUES ('NAN', 'NAN', 'NAN');
INSERT INTO t2 VALUES (8452345, 8452345, 8452345);
INSERT INTO t2 VALUES (164523.021354, 164523.021354, 164523.021354);
INSERT INTO t2 VALUES (0.02218383, 0.02218383, 0.02218383);
INSERT INTO t2 VALUES (-8452345, -8452345, -8452345);
INSERT INTO t2 VALUES (-164523.021354, -164523.021354, -164523.021354);
INSERT INTO t2 VALUES (-0.02218383, -0.02218383, -0.02218383);
INSERT INTO t2 VALUES ('NAN', 'NAN', 'NAN');
INSERT INTO t3 VALUES (0.002343544, 0.002343544, 0.002343544);
INSERT INTO t3 VALUES (0.00943244, 0.00943244, 0.00943244);
INSERT INTO t3 VALUES (-0.002343544, -0.002343544, -0.002343544);
INSERT INTO t3 VALUES (-0.00943244, -0.00943244, -0.00943244);
INSERT INTO t3 VALUES ('NAN', 'NAN', 'NAN');
SELECT * FROM t1;
SELECT * FROM t2;
SELECT * FROM t3;
UPDATE t1 SET a = 1999999, b = 1999999 where c = 'NAN';
UPDATE t2 SET a = 1999999, b = 1999999 where c = 'NAN';
UPDATE t3 SET a = 0.00123458345, b = 0.00395345 where c = 'NAN';
SELECT * FROM t1, t2, t3 where t1.c='NAN' and t2.c='NAN' and t3.c='NAN';
DELETE FROM t1;
DELETE FROM t2;
DELETE FROM t3;

-- test normal functions(invalid values)
INSERT INTO t1(a) VALUES (222222222.22222);
INSERT INTO t1(b) VALUES (222222222.22222);
INSERT INTO t1(c) VALUES (222222222.22222);
INSERT INTO t1(a) VALUES (9999599);
INSERT INTO t1(b) VALUES (9999959);
INSERT INTO t1(c) VALUES (9999995);
INSERT INTO t2(a) VALUES (222222222.22222);
INSERT INTO t2(b) VALUES (222222222.22222);
INSERT INTO t2(c) VALUES (222222222.22222);
INSERT INTO t2(a) VALUES (9995999);
INSERT INTO t2(b) VALUES (9959999);
INSERT INTO t2(c) VALUES (9599999);
INSERT INTO t3(a) VALUES (0.123234214);
INSERT INTO t3(b) VALUES (0.123234214);
INSERT INTO t3(c) VALUES (0.123234214);
INSERT INTO t3(a) VALUES (0.00999995);
INSERT INTO t3(b) VALUES (0.00995);
INSERT INTO t3(c) VALUES (0.0099999999995);
DROP TABLE t1;
DROP TABLE t2;
DROP TABLE t3;

-- boundary test
CREATE TABLE t1(a numeric(1,1000));
CREATE TABLE t2(a numeric(1,-84));
\d t1;
\d t2;
DROP TABLE t1;
DROP TABLE t2;

-- test alter table to  decrease scale
CREATE TABLE t1 (c1 int, c2 numeric(5, 2), c3 numeric(5, -2));
INSERT INTO t1 VALUES (1, 546.12, 456135.12);
SELECT * FROM t1;
ALTER TABLE t1 MODIFY (c2 numeric(5, 1)); -- error
ALTER TABLE t1 MODIFY (c3 numeric(5, -3)); -- error
SELECT * FROM t1;
DROP TABLE t1;

CREATE TABLE t3(a numeric(1,1001));
CREATE TABLE t3(a numeric(1,-85));
CREATE TABLE t3(a numeric(1,1001));
CREATE TABLE t3(a numeric(1,-85));
CREATE TABLE t3(a number(1,1001));
CREATE TABLE t3(a number(1,-85));
CREATE TABLE t3(a dec(1,1001));
CREATE TABLE t3(a dec(1,-85));
CREATE TABLE t3(a decimal(1,1001));
CREATE TABLE t3(a decimal(1,-85));
CREATE TABLE t3(a integer(1,1001));
CREATE TABLE t3(a integer(1,-85));
CREATE TABLE t3(a numeric(1,-32768));
CREATE TABLE t3(a numeric(1,32768));
CREATE TABLE t3(a number(1,-32768));
CREATE TABLE t3(a number(1,32768));
CREATE TABLE t3(a dec(1,-32768));
CREATE TABLE t3(a dec(1,32768));
CREATE TABLE t3(a decimal(1,-32768));
CREATE TABLE t3(a decimal(1,32768));
CREATE TABLE t3(a integer(1,-32768));
CREATE TABLE t3(a integer(1,32768));
-- PL/SQL test

CREATE OR REPLACE PACKAGE pak1 as 
var1 numeric(3,-4);
var2 numeric(5, 6);
type tp_tb1 is table of var1%type;
tb1 tp_tb1;
type tp_tb2 is table of var2%type;
tb2 tp_tb2;
procedure p1;
end pak1;
/
CREATE OR REPLACE package body pak1 as 
procedure p1 as
begin
tb1 = tp_tb1(1235234, 3241235.32432456, 0.00000002342, -1235234, -3241235.32432456, -0.00000002342);
raise info '%', tb1;
tb2 = tp_tb2(0.0123456878, 0.000565293244, -0.0123456878, -0.000565293244);
raise info '%', tb2;
end;
end pak1;
/
call pak1.p1();
DROP PACKAGE pak1;

reset behavior_compat_options;
reset current_schema;
drop schema numeric_negative_scale_test cascade;