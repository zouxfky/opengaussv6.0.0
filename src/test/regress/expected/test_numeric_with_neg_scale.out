create schema numeric_negative_scale_test;
set current_schema to numeric_negative_scale_test;
set behavior_compat_options = 'truncate_numeric_tail_zero';
-- test normal functions(valid values)
CREATE TABLE t1(a numeric(4,-3), b numeric(5,-2), c numeric(6, -1));
CREATE TABLE t2(a numeric(3,-4), b numeric(2,-5), c numeric(1, -6));
CREATE TABLE t3(a numeric(5, 7), b numeric(2, 4), c numeric(10, 12));
\d t1;
Table "numeric_negative_scale_test.t1"
 Column |     Type      | Modifiers 
--------+---------------+-----------
 a      | numeric(4,-3) | 
 b      | numeric(5,-2) | 
 c      | numeric(6,-1) | 

\d t2;
Table "numeric_negative_scale_test.t2"
 Column |     Type      | Modifiers 
--------+---------------+-----------
 a      | numeric(3,-4) | 
 b      | numeric(2,-5) | 
 c      | numeric(1,-6) | 

\d t3;
Table "numeric_negative_scale_test.t3"
 Column |      Type      | Modifiers 
--------+----------------+-----------
 a      | numeric(5,7)   | 
 b      | numeric(2,4)   | 
 c      | numeric(10,12) | 

INSERT INTO t1 VALUES (1234567, 1234567, 1234567);
INSERT INTO t1 VALUES (123.9435435, 123.9435435, 123.9435435);
INSERT INTO t1 VALUES (0.8293453, 0.8293453, 0.8293453);
INSERT INTO t1 VALUES (-1234567, -1234567, -1234567);
INSERT INTO t1 VALUES (-123.9435435, -123.9435435, -123.9435435);
INSERT INTO t1 VALUES (-0.8293453, -0.8293453, -0.8293453);
INSERT INTO t1 VALUES ('NAN', 'NAN', 'NAN');
INSERT INTO t2 VALUES (8452345, 8452345, 8452345);
INSERT INTO t2 VALUES (164523.021354, 164523.021354, 164523.021354);
INSERT INTO t2 VALUES (0.02218383, 0.02218383, 0.02218383);
INSERT INTO t2 VALUES (-8452345, -8452345, -8452345);
INSERT INTO t2 VALUES (-164523.021354, -164523.021354, -164523.021354);
INSERT INTO t2 VALUES (-0.02218383, -0.02218383, -0.02218383);
INSERT INTO t2 VALUES ('NAN', 'NAN', 'NAN');
INSERT INTO t3 VALUES (0.002343544, 0.002343544, 0.002343544);
INSERT INTO t3 VALUES (0.00943244, 0.00943244, 0.00943244);
INSERT INTO t3 VALUES (-0.002343544, -0.002343544, -0.002343544);
INSERT INTO t3 VALUES (-0.00943244, -0.00943244, -0.00943244);
INSERT INTO t3 VALUES ('NAN', 'NAN', 'NAN');
SELECT * FROM t1;
    a     |    b     |    c     
----------+----------+----------
  1235000 |  1234600 |  1234570
        0 |      100 |      120
        0 |        0 |        0
 -1235000 | -1234600 | -1234570
        0 |     -100 |     -120
        0 |        0 |        0
      NaN |      NaN |      NaN
(7 rows)

SELECT * FROM t2;
    a     |    b     |    c     
----------+----------+----------
  8450000 |  8500000 |  8000000
   160000 |   200000 |        0
        0 |        0 |        0
 -8450000 | -8500000 | -8000000
  -160000 |  -200000 |        0
        0 |        0 |        0
      NaN |      NaN |      NaN
(7 rows)

SELECT * FROM t3;
     a     |   b    |      c      
-----------+--------+-------------
  .0023435 |  .0023 |  .002343544
  .0094324 |  .0094 |   .00943244
 -.0023435 | -.0023 | -.002343544
 -.0094324 | -.0094 |  -.00943244
       NaN |    NaN |         NaN
(5 rows)

UPDATE t1 SET a = 1999999, b = 1999999 where c = 'NAN';
UPDATE t2 SET a = 1999999, b = 1999999 where c = 'NAN';
UPDATE t3 SET a = 0.00123458345, b = 0.00395345 where c = 'NAN';
SELECT * FROM t1, t2, t3 where t1.c='NAN' and t2.c='NAN' and t3.c='NAN';
    a    |    b    |  c  |    a    |    b    |  c  |    a     |  b   |  c  
---------+---------+-----+---------+---------+-----+----------+------+-----
 2000000 | 2000000 | NaN | 2000000 | 2000000 | NaN | .0012346 | .004 | NaN
(1 row)

DELETE FROM t1;
DELETE FROM t2;
DELETE FROM t3;
-- test normal functions(invalid values)
INSERT INTO t1(a) VALUES (222222222.22222);
ERROR:  numeric field overflow
DETAIL:  A field with precision 4, scale -3 must round to an absolute value less than 10^7.
CONTEXT:  referenced column: a
INSERT INTO t1(b) VALUES (222222222.22222);
ERROR:  numeric field overflow
DETAIL:  A field with precision 5, scale -2 must round to an absolute value less than 10^7.
CONTEXT:  referenced column: b
INSERT INTO t1(c) VALUES (222222222.22222);
ERROR:  numeric field overflow
DETAIL:  A field with precision 6, scale -1 must round to an absolute value less than 10^7.
CONTEXT:  referenced column: c
INSERT INTO t1(a) VALUES (9999599);
ERROR:  numeric field overflow
DETAIL:  A field with precision 4, scale -3 must round to an absolute value less than 10^7.
CONTEXT:  referenced column: a
INSERT INTO t1(b) VALUES (9999959);
ERROR:  numeric field overflow
DETAIL:  A field with precision 5, scale -2 must round to an absolute value less than 10^7.
CONTEXT:  referenced column: b
INSERT INTO t1(c) VALUES (9999995);
ERROR:  numeric field overflow
DETAIL:  A field with precision 6, scale -1 must round to an absolute value less than 10^7.
CONTEXT:  referenced column: c
INSERT INTO t2(a) VALUES (222222222.22222);
ERROR:  numeric field overflow
DETAIL:  A field with precision 3, scale -4 must round to an absolute value less than 10^7.
CONTEXT:  referenced column: a
INSERT INTO t2(b) VALUES (222222222.22222);
ERROR:  numeric field overflow
DETAIL:  A field with precision 2, scale -5 must round to an absolute value less than 10^7.
CONTEXT:  referenced column: b
INSERT INTO t2(c) VALUES (222222222.22222);
ERROR:  numeric field overflow
DETAIL:  A field with precision 1, scale -6 must round to an absolute value less than 10^7.
CONTEXT:  referenced column: c
INSERT INTO t2(a) VALUES (9995999);
ERROR:  numeric field overflow
DETAIL:  A field with precision 3, scale -4 must round to an absolute value less than 10^7.
CONTEXT:  referenced column: a
INSERT INTO t2(b) VALUES (9959999);
ERROR:  numeric field overflow
DETAIL:  A field with precision 2, scale -5 must round to an absolute value less than 10^7.
CONTEXT:  referenced column: b
INSERT INTO t2(c) VALUES (9599999);
ERROR:  numeric field overflow
DETAIL:  A field with precision 1, scale -6 must round to an absolute value less than 10^7.
CONTEXT:  referenced column: c
INSERT INTO t3(a) VALUES (0.123234214);
ERROR:  numeric field overflow
DETAIL:  A field with precision 5, scale 7 must round to an absolute value less than 10^-2.
CONTEXT:  referenced column: a
INSERT INTO t3(b) VALUES (0.123234214);
ERROR:  numeric field overflow
DETAIL:  A field with precision 2, scale 4 must round to an absolute value less than 10^-2.
CONTEXT:  referenced column: b
INSERT INTO t3(c) VALUES (0.123234214);
ERROR:  numeric field overflow
DETAIL:  A field with precision 10, scale 12 must round to an absolute value less than 10^-2.
CONTEXT:  referenced column: c
INSERT INTO t3(a) VALUES (0.00999995);
ERROR:  numeric field overflow
DETAIL:  A field with precision 5, scale 7 must round to an absolute value less than 10^-2.
CONTEXT:  referenced column: a
INSERT INTO t3(b) VALUES (0.00995);
ERROR:  numeric field overflow
DETAIL:  A field with precision 2, scale 4 must round to an absolute value less than 10^-2.
CONTEXT:  referenced column: b
INSERT INTO t3(c) VALUES (0.0099999999995);
ERROR:  numeric field overflow
DETAIL:  A field with precision 10, scale 12 must round to an absolute value less than 10^-2.
CONTEXT:  referenced column: c
DROP TABLE t1;
DROP TABLE t2;
DROP TABLE t3;
-- boundary test
CREATE TABLE t1(a numeric(1,1000));
CREATE TABLE t2(a numeric(1,-84));
\d t1;
Table "numeric_negative_scale_test.t1"
 Column |      Type       | Modifiers 
--------+-----------------+-----------
 a      | numeric(1,1000) | 

\d t2;
Table "numeric_negative_scale_test.t2"
 Column |      Type      | Modifiers 
--------+----------------+-----------
 a      | numeric(1,-84) | 

DROP TABLE t1;
DROP TABLE t2;
-- test alter table to  decrease scale
CREATE TABLE t1 (c1 int, c2 numeric(5, 2), c3 numeric(5, -2));
INSERT INTO t1 VALUES (1, 546.12, 456135.12);
SELECT * FROM t1;
 c1 |   c2   |   c3   
----+--------+--------
  1 | 546.12 | 456100
(1 row)

ALTER TABLE t1 MODIFY (c2 numeric(5, 1)); -- error
ERROR:  column to be modified must be empty to decrease precision or scale
ALTER TABLE t1 MODIFY (c3 numeric(5, -3)); -- error
ERROR:  column to be modified must be empty to decrease precision or scale
SELECT * FROM t1;
 c1 |   c2   |   c3   
----+--------+--------
  1 | 546.12 | 456100
(1 row)

DROP TABLE t1;
CREATE TABLE t3(a numeric(1,1001));
ERROR:  NUMERIC scale must be between -84 and 1000
LINE 1: CREATE TABLE t3(a numeric(1,1001));
                                 ^
CREATE TABLE t3(a numeric(1,-85));
ERROR:  NUMERIC scale must be between -84 and 1000
LINE 1: CREATE TABLE t3(a numeric(1,-85));
                                 ^
CREATE TABLE t3(a numeric(1,1001));
ERROR:  NUMERIC scale must be between -84 and 1000
LINE 1: CREATE TABLE t3(a numeric(1,1001));
                                 ^
CREATE TABLE t3(a numeric(1,-85));
ERROR:  NUMERIC scale must be between -84 and 1000
LINE 1: CREATE TABLE t3(a numeric(1,-85));
                                 ^
CREATE TABLE t3(a number(1,1001));
ERROR:  NUMERIC scale must be between -84 and 1000
LINE 1: CREATE TABLE t3(a number(1,1001));
                                ^
CREATE TABLE t3(a number(1,-85));
ERROR:  NUMERIC scale must be between -84 and 1000
LINE 1: CREATE TABLE t3(a number(1,-85));
                                ^
CREATE TABLE t3(a dec(1,1001));
ERROR:  NUMERIC scale must be between -84 and 1000
LINE 1: CREATE TABLE t3(a dec(1,1001));
                             ^
CREATE TABLE t3(a dec(1,-85));
ERROR:  NUMERIC scale must be between -84 and 1000
LINE 1: CREATE TABLE t3(a dec(1,-85));
                             ^
CREATE TABLE t3(a decimal(1,1001));
ERROR:  NUMERIC scale must be between -84 and 1000
LINE 1: CREATE TABLE t3(a decimal(1,1001));
                                 ^
CREATE TABLE t3(a decimal(1,-85));
ERROR:  NUMERIC scale must be between -84 and 1000
LINE 1: CREATE TABLE t3(a decimal(1,-85));
                                 ^
CREATE TABLE t3(a integer(1,1001));
ERROR:  NUMERIC scale must be between -84 and 1000
LINE 1: CREATE TABLE t3(a integer(1,1001));
                                 ^
CREATE TABLE t3(a integer(1,-85));
ERROR:  NUMERIC scale must be between -84 and 1000
LINE 1: CREATE TABLE t3(a integer(1,-85));
                                 ^
CREATE TABLE t3(a numeric(1,-32768));
ERROR:  NUMERIC scale must be between -84 and 1000
LINE 1: CREATE TABLE t3(a numeric(1,-32768));
                                 ^
CREATE TABLE t3(a numeric(1,32768));
ERROR:  NUMERIC scale must be between -84 and 1000
LINE 1: CREATE TABLE t3(a numeric(1,32768));
                                 ^
CREATE TABLE t3(a number(1,-32768));
ERROR:  NUMERIC scale must be between -84 and 1000
LINE 1: CREATE TABLE t3(a number(1,-32768));
                                ^
CREATE TABLE t3(a number(1,32768));
ERROR:  NUMERIC scale must be between -84 and 1000
LINE 1: CREATE TABLE t3(a number(1,32768));
                                ^
CREATE TABLE t3(a dec(1,-32768));
ERROR:  NUMERIC scale must be between -84 and 1000
LINE 1: CREATE TABLE t3(a dec(1,-32768));
                             ^
CREATE TABLE t3(a dec(1,32768));
ERROR:  NUMERIC scale must be between -84 and 1000
LINE 1: CREATE TABLE t3(a dec(1,32768));
                             ^
CREATE TABLE t3(a decimal(1,-32768));
ERROR:  NUMERIC scale must be between -84 and 1000
LINE 1: CREATE TABLE t3(a decimal(1,-32768));
                                 ^
CREATE TABLE t3(a decimal(1,32768));
ERROR:  NUMERIC scale must be between -84 and 1000
LINE 1: CREATE TABLE t3(a decimal(1,32768));
                                 ^
CREATE TABLE t3(a integer(1,-32768));
ERROR:  NUMERIC scale must be between -84 and 1000
LINE 1: CREATE TABLE t3(a integer(1,-32768));
                                 ^
CREATE TABLE t3(a integer(1,32768));
ERROR:  NUMERIC scale must be between -84 and 1000
LINE 1: CREATE TABLE t3(a integer(1,32768));
                                 ^
-- PL/SQL test
CREATE OR REPLACE PACKAGE pak1 as 
var1 numeric(3,-4);
var2 numeric(5, 6);
type tp_tb1 is table of var1%type;
tb1 tp_tb1;
type tp_tb2 is table of var2%type;
tb2 tp_tb2;
procedure p1;
end pak1;
/
CREATE OR REPLACE package body pak1 as 
procedure p1 as
begin
tb1 = tp_tb1(1235234, 3241235.32432456, 0.00000002342, -1235234, -3241235.32432456, -0.00000002342);
raise info '%', tb1;
tb2 = tp_tb2(0.0123456878, 0.000565293244, -0.0123456878, -0.000565293244);
raise info '%', tb2;
end;
end pak1;
/
call pak1.p1();
INFO:  {1240000,3240000,0,-1240000,-3240000,0}
INFO:  {.012346,.000565,-.012346,-.000565}
 p1 
----
 
(1 row)

DROP PACKAGE pak1;
NOTICE:  drop cascades to function numeric_negative_scale_test.p1()
reset behavior_compat_options;
reset current_schema;
drop schema numeric_negative_scale_test cascade;
